File: packages/python/qsu/object/objUpdate.py
from ..verify.isObject import isObject


def objUpdate(obj, key, value, recursive=False, upsert=False):
	if not isinstance(obj, dict):
		return None

	hasUpdated = {'value': False}

	# Work on copies. `newObj = obj` modified the caller's dict (and every nested dict)
	# in place, even though the function looks like it returns a new one.
	def updateObject(currentObj):
		result = dict(currentObj)

		for currentKey in list(result.keys()):
			if recursive and result.get(currentKey) and isObject(result[currentKey]):
				result[currentKey] = updateObject(result[currentKey])

		# Assign once per dict. The old code repeated this identical assignment for
		# every key in the dict.
		if key in result:
			result[key] = value
			hasUpdated['value'] = True

		return result

	newObj = updateObject(obj)

	if not hasUpdated['value'] and upsert:
		newObj[key] = value

	return newObj

File: packages/javascript/lib/object/objUpdate.ts
import { AnyValueObject } from '../_types/global';
import { isObject } from '../verify/isObject.js';

export function objUpdate(
	obj: AnyValueObject,
	key: string,
	value: any,
	recursive = false,
	upsert = false
): AnyValueObject | null {
	if (!obj || typeof obj !== 'object') {
		return null;
	}

	let hasUpdated = false;

	// Work on copies. `Object.assign(obj, {})` returned `obj` itself, so the caller's object
	// (and every nested object) was modified in place.
	const updateObject = (currentObj: AnyValueObject): AnyValueObject => {
		const result: AnyValueObject = { ...currentObj };
		// Build the key list once. Calling `Object.keys` inside the loop rebuilt the whole
		// array on every iteration, making this O(n^2).
		const keys = Object.keys(result);

		for (let i = 0, keyLength = keys.length; i < keyLength; i += 1) {
			const currentKey = keys[i];

			if (recursive && result[currentKey] && isObject(result[currentKey])) {
				result[currentKey] = updateObject(result[currentKey]);
			}
		}

		// Assign once per object. The old code repeated this identical assignment for every
		// key in the object.
		if (Object.hasOwn(result, key)) {
			result[key] = value;
			hasUpdated = true;
		}

		return result;
	};

	const newObj = updateObject(obj);

	if (!hasUpdated && upsert) {
		newObj[key] = value;
	}

	return newObj;
}
